sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -517,3 +519,18 @@ case class ScalaAggregator[IN, BUF, OUT](
 
   override def nodeName: String = agg.getClass.getSimpleName
 }
+
+/**
+ * An extension rule to resolve encoder expressions from a [[ScalaAggregator]]
+ */
+object ResolveEncodersInScalaAgg extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+    case p if !p.resolved => p
+    case p => p.transformExpressionsUp {
+      case agg: ScalaAggregator[_, _, _] =>
+        agg.copy(
+          inputEncoder = agg.inputEncoder.resolveAndBind(),
```

Comment:
A followup we can do is to resolve and bind using the actual input data types, so that we can do casting or reorder fields.

Reply:
That would be nice.  I tried this and but the way I did it wasn't having any effect.

Reply:
@cloud-fan  what I had done earlier was:
```scala
object ResolveEncodersInScalaAgg extends Rule[LogicalPlan] {
  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
    case p if !p.resolved => p
    case p => p.transformExpressionsUp {
      case agg: ScalaAggregator[_, _, _] =>
        val children = agg.children
        require(children.length > 0, "Missing aggregator input")
        val dataType: DataType = if (children.length == 1) children.head.dataType else {
          StructType(children.map(_.dataType).zipWithIndex.map { case (dt, j) =>
            StructField(s"_$j", dt, true)
          })
        }
        val attrs = if (agg.inputEncoder.isSerializedAsStructForTopLevel) {
          dataType.asInstanceOf[StructType].toAttributes
        } else {
          (new StructType().add("input", dataType)).toAttributes
        }
        agg.copy(
          inputEncoder = agg.inputEncoder.resolveAndBind(attrs),
          bufferEncoder = agg.bufferEncoder.resolveAndBind())
    }
  }
}
```
This also passes unit tests, but it would still fail if I tried to give it `Float` data, so it's not automatically casting.